File: leetcode/algorithms/39_combination_sum/main.js
/**
 * Backtracking
 *
 * Complexities:
 *   N - Size of `candidates`
 *   T - Value of `target`
 *   M - Minimum value of `candidates`
 *   - Time Complexity: O(Nᵀ/ᴹ)
 *   - Space Complexity: O(T/M)
 */
/**
 * @param {number[]} candidates
 * @param {number} target
 * @param {number} start
 * @param {number[]} current
 * @param {number[][]} result
 */
var backtracking = function (candidates, target, start, current, result) {
  if (target == 0) {
    result.push([...current]);
    return;
  }
  if (target < 0) {
    return;
  }

  for (let i = start; i < candidates.length; i++) {
    current.push(candidates[i]);

    backtracking(candidates, target - candidates[i], i, current, result);

    current.pop();
  }
};

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function (candidates, target) {
  const result = new Array();
  const current = new Array();

  backtracking(candidates, target, 0, current, result);

  return result;
};


// Solution
/**
 * Backtracking
 *
 * Complexities:
 *   N - Size of `candidates`
 *   T - Value of `target`
 *   M - Minimum value of `candidates`
 *   - Time Complexity: O(Nᵀ/ᴹ)
 *   - Space Complexity: O(T/M)
 */
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var solution = function (candidates, target) {
  const result = [];

  candidates.sort((a, b) => a - b);

  /**
   * @param {number} remain
   * @param {number[]} currentComb
   * @param {number} start
   */
  function backtrack(remain, currentComb, start) {
    if (remain === 0) {
      result.push([...currentComb]);
      return;
    }

    for (let i = start; i < candidates.length; i++) {
      const currentNum = candidates[i];

      if (currentNum > remain) {
        break;
      }

      currentComb.push(currentNum);

      backtrack(remain - currentNum, currentComb, i);

      currentComb.pop();
    }
  }

  backtrack(target, [], 0);

  return result;
};

File: leetcode/algorithms/39_combination_sum/main.py
from typing import List


class CombinationSum:
    """
    # Backtracking
    # 
    # Complexities:
    #   N - Size of `candidates`
    #   T - Value of `target`
    #   M - Minimum value of `candidates`
    #   - Time Complexity: O(Nᵀ/ᴹ)
    #   - Space Complexity: O(T/M)
    """
    def backtracking(self, candidates: List[int], target: int, start: int, current: List[int], result: List[List[int]]) -> None:
        if target == 0:
            result.append(current[:])
            return

        if target < 0:
            return

        for i in range(start, len(candidates)):
            current.append(candidates[i])

            self.backtracking(candidates, target -
                              candidates[i], i, current, result)

            current.pop()

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result: List[List[int]] = []
        current: List[int] = []

        self.backtracking(candidates, target, 0, current, result)

        return result


    # Solution
    """
    # Backtracking
    # 
    # Complexities:
    #   N - Size of `candidates`
    #   T - Value of `target`
    #   M - Minimum value of `candidates`
    #   - Time Complexity: O(Nᵀ/ᴹ)
    #   - Space Complexity: O(T/M)
    """
    def solution(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []

        candidates.sort()

        def backtrack(remain: int, curr_path: List[int], start: int):
            if remain == 0:
                result.append(curr_path[:])
                return

            for i in range(start, len(candidates)):
                if remain - candidates[i] < 0:
                    break

                curr_path.append(candidates[i])

                backtrack(remain - candidates[i], curr_path, i)

                curr_path.pop()

        backtrack(target, [], 0)

        return result
